### PR TITLE
TPI.py First Attempt

### PR DESCRIPTION
**Pull request by miltonash**
Here is my first try at TPI.py. Atom's linter is suggesting some more things need to be initialized in `params`. Please take a look and give me your feedback.

### CODE REVIEW

File: scripts/TPI.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    while dist > mindist and tpi_iter < maxiter:
+
+        # Define paths
+        b_11 = 1.1 * b_ss
+        BQ_params = (g_n_path[0], omega_S_preTP, rho_s)
+        K_params = (g_n_path[0], omega_S_preTP, imm_rates_path[0, :])
+        BQ_11 = agg.get_BQ(b_11, r_11, BQ_params)
+        K_11 = agg.get_K(b_11, K_params)
+        BQpath_init = np.zeros(T + S - 1)
+        Kpath_init = np.zeros(T + S - 1)
```
Comment (jdebacker) [inline scripts/TPI.py]: Don't need to compute `Kpath` here.

File: scripts/TPI.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                omega_path_S[:T - 1, 1:] * bmat[:, 1:T].T).sum(axis=1)
+        new_BQpath = np.zeros(T)
+        new_BQpath[0] = BQ_1
+        new_BQpath[1:] = \
+            ((1 + rpath[1:T]) / (rho_s[:-1]) *
+                omega_path_S[:T - 1, :-1] * bmat[:, 1:T].T).sum(axis=1)
+
+        KBQ_init = np.append(Kpath_init[:T], BQpath_init[:T])
+        Kpath_init[:T] = xi * new_Kpath[:T] + (1 - xi) * Kpath_init[:T]
+        new_KBQ = np.append(new_Kpath[:T], new_BQpath[:T])
```
Comment (jdebacker) [inline scripts/TPI.py]: "Loop will only be over guesses of `r_path` and `BQ_path`, so these are the only things you need to compute new values of and the distance with.  See Issue #23.

**Comment by jdebacker** [comment]
This looks pretty good.  I had a couple comments.  Address those and then let me know when you think this is ready to merge.

File: scripts/TPI.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        BQpath_init[:T] = xi * new_BQpath[:T] + (1 - xi) * BQpath_init[:T]
+        # update iteration counter
+        tpi_iter += 1
+
+    if tpi_iter < maxiter:
+        print('The time path solved! ->', ' iter:', tpi_iter, ', dist: ', dist)
+    else:
+        print('The time path did not solve.')
+
+    return new_KBQ[:T]
```
Comment (saharnazb) [inline scripts/TPI.py]: @miltonash Can you separate `Kpath` and `BQpath` in the return or @jdebacker I should use `.get(new_BQpath)` and `.get(new_Kpath)` in execute function after `solve_tp()`. I doubt if `.get()` will work. I want to plot their path in `execute.py`.

Comment (jdebacker) [reply scripts/TPI.py]: I agree with @saharnazb that you should return the results in a way that will be easier to use later (e.g. in a tuple (or, better, a dictionary)), with variable of interest as a separate element.

File: scripts/TPI.py (hunk 2 above)
Comment (saharnazb) [inline scripts/TPI.py]: I think you can remove this line. You don't need to append two paths.  (Line 77)

And in the last line:
`return [new_Kpath, new_BQpath]`
This way I can add the following to execute:
`new_Kpath, _ = solve_tp("arguments")`
`_, new_BQpath = solve_tp("arguments")`
Please confirm or let me know if any other ideas, any of you have.

Comment (LeiliPR) [reply scripts/TPI.py]: That is correct, as we dont need kpath. @miltonash please go ahead and change this part.

**Comment by miltonash** [comment]
See my commit message for changes. Did I get everything you guys pointed out? Look good?

**Comment by jdebacker** [comment]
I'll go ahead and merge this to get `TPI.py` into this repo.  But I do see some issues to address.  I'll post them in an issue and they can be resolved in future PRs.

Thanks for the work on this @miltonash !

**Pull request merged**